File: georegistry-server/src/main/java/net/geoprism/registry/service/ClassificationService.java
/**
 * Copyright (c) 2022 TerraFrame, Inc. All rights reserved.
 *
 * This file is part of Geoprism Registry(tm).
 *
 * Geoprism Registry(tm) is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Geoprism Registry(tm) is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Geoprism Registry(tm).  If not, see <http://www.gnu.org/licenses/>.
 */
package net.geoprism.registry.service;

import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collector;

import org.springframework.stereotype.Component;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.runwaysdk.session.Request;
import com.runwaysdk.session.RequestType;

import net.geoprism.registry.model.Classification;
import net.geoprism.registry.model.ClassificationType;
import net.geoprism.registry.view.Page;

@Component
public class ClassificationService
{
  @Request(RequestType.SESSION)
  public JsonObject apply(String sessionId, String classificationCode, String parentCode, JsonObject object, boolean isNew)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification parent = parentCode != null ? Classification.get(type, parentCode) : null;

    Classification classification = Classification.construct(type, object, isNew);
    classification.populate(object);
    classification.apply(parent);

    // Return the refreshed copy of the geoObject
    return classification.toJSON();
  }

  @Request(RequestType.SESSION)
  public void remove(String sessionId, String classificationCode, String code)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification classification = Classification.get(type, code);
    classification.delete();
  }

  @Request(RequestType.SESSION)
  public JsonObject get(String sessionId, String classificationCode, String code)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification classification = Classification.get(type, code);

    return classification.toJSON();
  }

  @Request(RequestType.SESSION)
  public void move(String sessionId, String classificationCode, String code, String parentCode)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);
    Classification classification = Classification.get(type, code);
    Classification newParent = Classification.get(type, parentCode);

    classification.move(newParent);
  }

  @Request(RequestType.SESSION)
  public void addChild(String sessionId, String classificationCode, String parentCode, String childCode)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification parent = Classification.get(type, parentCode);
    Classification child = Classification.get(type, childCode);

    parent.addChild(child);
  }

  @Request(RequestType.SESSION)
  public void removeChild(String sessionId, String classificationCode, String parentCode, String childCode)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification parent = Classification.get(type, parentCode);
    Classification child = Classification.get(type, childCode);

    parent.removeChild(child);
  }

  @Request(RequestType.SESSION)
  public JsonObject getChildren(String sessionId, String classificationCode, String code, Integer pageSize, Integer pageNumber)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    if (code != null)
    {
      Classification parent = Classification.get(type, code);

      return parent.getChildren(pageSize, pageNumber).toJSON();
    }

    Classification root = type.getRoot();
    List<Classification> roots = new LinkedList<Classification>();

    if (root != null)
    {
      roots.add(root);
    }

    return new Page<Classification>(roots.size(), pageNumber, pageSize, roots).toJSON();
  }

  @Request(RequestType.SESSION)
  public JsonObject getAncestorTree(String sessionId, String classificationCode, String rootCode, String code, Integer pageSize)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);

    Classification child = Classification.get(type, code);

    return child.getAncestorTree(rootCode, pageSize).toJSON();
  }

  @Request(RequestType.SESSION)
  public JsonArray search(String sessionId, String classificationCode, String rootCode, String text)
  {
    ClassificationType type = ClassificationType.getByCode(classificationCode);
    List<Classification> results = Classification.search(type, rootCode, text);

    return results.stream().map(child -> child.toJSON()).collect(Collector.of(() -> new JsonArray(), (r, t) -> r.add((JsonObject) t), (x1, x2) -> {
      x1.addAll(x2);
      return x1;
    }));
  }

}

File: georegistry-ui/src/app/registry/service/classification.service.ts
///
/// Copyright (c) 2022 TerraFrame, Inc. All rights reserved.
///
/// This file is part of Geoprism Registry(tm).
///
/// Geoprism Registry(tm) is free software: you can redistribute it and/or modify
/// it under the terms of the GNU Lesser General Public License as
/// published by the Free Software Foundation, either version 3 of the
/// License, or (at your option) any later version.
///
/// Geoprism Registry(tm) is distributed in the hope that it will be useful, but
/// WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU Lesser General Public License for more details.
///
/// You should have received a copy of the GNU Lesser General Public
/// License along with Geoprism Registry(tm).  If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from "@angular/core";
import { HttpClient, HttpHeaders, HttpParams } from "@angular/common/http";
import { finalize } from "rxjs/operators";

import { EventService } from "@shared/service";

import { GeoRegistryConfiguration } from "@core/model/core";
import { Classification, ClassificationNode } from "@registry/model/classification-type";
import { PageResult } from "@shared/model/core";

import { environment } from 'src/environments/environment';

@Injectable()
export class ClassificationService {

    // eslint-disable-next-line no-useless-constructor
    constructor(private http: HttpClient, private eventService: EventService) { }

    apply(classificationCode: string, parentCode: string, classification: Classification, isNew: boolean): Promise<Classification> {
        let headers = new HttpHeaders({
            "Content-Type": "application/json"
        });

        const params = {
            classificationCode: classificationCode,
            classification: classification,
            isNew: isNew
        };

        if (parentCode != null) {
            params["parentCode"] = parentCode;
        }

        this.eventService.start();

        return this.http
            .post<Classification>(environment.apiUrl + "/api/classification/apply", JSON.stringify(params), { headers: headers })
            .pipe(finalize(() => {
                this.eventService.complete();
            }))
            .toPromise();
    }

    remove(classificationCode: string, code: string): Promise<void> {
        let headers = new HttpHeaders({
            "Content-Type": "application/json"
        });

        const params = {
            classificationCode: classificationCode,
            code: code
        };

        this.eventService.start();

        return this.http
            .post<void>(environment.apiUrl + "/api/classification/remove", JSON.stringify(params), { headers: headers })
            .pipe(finalize(() => {
                this.eventService.complete();
            }))
            .toPromise();
    }

    move(classificationCode: string, code: string, parentCode: string): Promise<void> {
        let headers = new HttpHeaders({
            "Content-Type": "application/json"
        });

        const params = {
            classificationCode: classificationCode,
            code: code,
            parentCode: parentCode
        };

        this.eventService.start();

        return this.http
            .post<void>(environment.apiUrl + "/api/classification/move", JSON.stringify(params), { headers: headers })
            .pipe(finalize(() => {
                this.eventService.complete();
            }))
            .toPromise();
    }

    getChildren(classificationCode: string, code: string, pageNumber: number, pageSize: number): Promise<PageResult<Classification>> {
        let params: HttpParams = new HttpParams();
        params = params.set("classificationCode", classificationCode);
        params = params.set("pageNumber", pageNumber.toString());
        params = params.set("pageSize", pageSize.toString());

        if (code != null) {
            params = params.set("code", code);
        }

        return this.http.get<PageResult<Classification>>(environment.apiUrl + "/api/classification/get-children", { params: params })
            .toPromise();
    }

    getAncestorTree(classificationCode: string, rootCode: string, code: string, pageSize: number): Promise<ClassificationNode> {
        let params: HttpParams = new HttpParams();
        params = params.set("classificationCode", classificationCode);
        params = params.set("code", code);
        params = params.set("pageSize", pageSize.toString());

        if (rootCode != null) {
            params = params.set("rootCode", rootCode);
        }

        return this.http.get<ClassificationNode>(environment.apiUrl + "/api/classification/get-ancestor-tree", { params: params })
            .toPromise();
    }

    search(classificationCode: string, rootCode: string, text: string): Promise<Classification[]> {
        let params: HttpParams = new HttpParams();
        params = params.set("classificationCode", classificationCode);

        if (rootCode != null) {
            params = params.set("rootCode", rootCode);
        }

        if (text != null) {
            params = params.set("text", text);
        }

        return this.http.get<Classification[]>(environment.apiUrl + "/api/classification/search", { params: params })
            .toPromise();
    }

    get(classificationCode: string, code: string): Promise<Classification> {
        let params: HttpParams = new HttpParams();
        params = params.set("classificationCode", classificationCode);
        params = params.set("code", code);

        return this.http.get<Classification>(environment.apiUrl + "/api/classification/get", { params: params })
            .toPromise();
    }

}
